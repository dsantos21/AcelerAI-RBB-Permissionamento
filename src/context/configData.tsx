import React, { createContext, useContext } from 'react';
import { Config } from '../util/configLoader';

type ContextType =
  | {
      config: Config;
    }
  | undefined;

const ConfigDataContext = createContext<ContextType>(undefined);

export const ConfigDataProvider: React.FC<{ config: Config }> = props => {
  return <ConfigDataContext.Provider value={{ config: props.config }} {...props} />;
};

export const useConfig = () => {
  const context = useContext(ConfigDataContext);

  if (!context) {
    throw new Error('useConfig must be used within a ConfigDataProvider');
  }

  return context.config;
};
