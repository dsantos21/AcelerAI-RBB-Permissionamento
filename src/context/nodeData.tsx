// Libs
import React, { createContext, useContext, useEffect, useState, useMemo } from 'react';
import { NodeRules } from '../chain/@types/NodeRules';
import { nodeRulesFactory } from '../chain/contracts/NodeRules';
import { useNetwork } from './network';

// Utils
import { paramsToIdentifier, Enode as RawEnode } from '../util/enodetools';

type Enode = RawEnode & { identifier: string };

type ContextType =
  | {
      nodeList: Enode[];
      setNodeList: React.Dispatch<React.SetStateAction<Enode[]>>;
      nodeReadOnly?: boolean;
      setNodeReadOnly: React.Dispatch<React.SetStateAction<boolean | undefined>>;
      nodeRulesContract?: NodeRules;
      setNodeRulesContract: React.Dispatch<React.SetStateAction<NodeRules | undefined>>;
    }
  | undefined;

const DataContext = createContext<ContextType>(undefined);

const loadNodeData = (
  nodeRulesContract: NodeRules | undefined,
  setNodeList: (account: Enode[]) => void,
  setNodeReadOnly: (readOnly?: boolean) => void
) => {
  if (nodeRulesContract === undefined) {
    setNodeList([]);
    setNodeReadOnly(undefined);
  } else {
    nodeRulesContract.functions.isReadOnly().then(isReadOnly => setNodeReadOnly(isReadOnly));
    nodeRulesContract.functions.getSize().then(listSize => {
      const listElementPromises = [];
      for (let i = 0; listSize.gt(i); i++) {
        listElementPromises.push(nodeRulesContract.functions.getByIndex(i));
      }
      Promise.all(listElementPromises).then(responses => {
        const updatedNodeList = responses.map(r => {
          const withStringyPort = { ...r, port: r.port.toString() };
          return {
            ...withStringyPort,
            identifier: paramsToIdentifier(withStringyPort)
          };
        });
        setNodeList(updatedNodeList);
      });
    });
  }
};

/**
 * Provider for the data context that contains the node list
 * @param {Object} props Props given to the NodeDataProvider
 * @return The provider with the following value:
 *  - nodeList: list of whiteliist enode from Node Rules contract
 *  - setNodeList: setter for the list state
 */
export const NodeDataProvider: React.FC<{}> = props => {
  const [nodeList, setNodeList] = useState<Enode[]>([]);
  const [nodeReadOnly, setNodeReadOnly] = useState<boolean | undefined>(undefined);
  const [nodeRulesContract, setNodeRulesContract] = useState<NodeRules | undefined>(undefined);

  const value = useMemo(
    () => ({ nodeList, setNodeList, nodeReadOnly, setNodeReadOnly, nodeRulesContract, setNodeRulesContract }),
    [nodeList, setNodeList, nodeReadOnly, setNodeReadOnly, nodeRulesContract, setNodeRulesContract]
  );

  const { nodeIngressContract } = useNetwork();

  useEffect(() => {
    if (nodeIngressContract === undefined) {
      setNodeRulesContract(undefined);
    } else {
      nodeRulesFactory(nodeIngressContract).then(contract => {
        setNodeRulesContract(contract);
        contract.removeAllListeners('NodeAdded');
        contract.removeAllListeners('NodeRemoved');
        contract.on('NodeAdded', () => {
          loadNodeData(contract, setNodeList, setNodeReadOnly);
        });
        contract.on('NodeRemoved', () => {
          loadNodeData(contract, setNodeList, setNodeReadOnly);
        });
      });
    }
  }, [nodeIngressContract]);

  return <DataContext.Provider value={value} {...props} />;
};

/**
 * Fetch the appropriate node data on chain and synchronize with it
 * @return {Object} Contains data of interest:
 *  - dataReady: true if isReadOnly and node allowlist are correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isReadOnly: Node contract is lock or unlock,
 *  - allowlist: list of permitted nodes from Node contract,
 */
export const useNodeData = () => {
  const context = useContext(DataContext);
  if (!context) {
    throw new Error('useNodeData must be used within a NodeDataProvider.');
  }

  const { nodeList, setNodeList, nodeReadOnly, setNodeReadOnly, nodeRulesContract } = context;

  useEffect(() => {
    loadNodeData(nodeRulesContract, setNodeList, setNodeReadOnly);
  }, [nodeRulesContract, setNodeList, setNodeReadOnly]);

  const formattedNodeList = useMemo(() => {
    return nodeList.map(enode => ({ ...enode, status: 'active' })).reverse();
  }, [nodeList]);

  const dataReady = useMemo(() => {
    return nodeRulesContract !== undefined && nodeReadOnly !== undefined && nodeList !== undefined;
  }, [nodeRulesContract, nodeReadOnly, nodeList]);

  return {
    dataReady,
    allowlist: formattedNodeList,
    isReadOnly: nodeReadOnly,
    nodeRulesContract
  };
};
