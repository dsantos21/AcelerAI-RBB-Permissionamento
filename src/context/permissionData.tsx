// Libs
import React, { createContext, useContext, useEffect, useState, useMemo } from 'react';
import { AccountRules } from '../chain/@types/AccountRules';
import { accountRulesFactory } from '../chain/contracts/AccountRules';
import { useNetwork } from './network';

export type Permission = {
  type: string;
  authorizer: string;
  authorized: string;
  date: string;
  txHash: string;
  identifier: string;
};

export type PermissionFilter = {
  accountAdded: boolean | null;
  accountGrantee: string | null;
  accountGrantor: string | null;
  blockTimestamp: Date | null;
};

type ContextType =
  | {
      permissionList: Permission[];
      setPermissionList: React.Dispatch<React.SetStateAction<Permission[]>>;
      accountReadOnly?: boolean;
      setAccountReadOnly: React.Dispatch<React.SetStateAction<boolean | undefined>>;
      permissionRulesContract?: AccountRules;
      setAccountRulesContract: React.Dispatch<React.SetStateAction<AccountRules | undefined>>;
      permissionFilter?: PermissionFilter;
      setPermissionFilter: React.Dispatch<React.SetStateAction<PermissionFilter>>;
    }
  | undefined;

const DataContext = createContext<ContextType>(undefined);

const loadPermissionData = (
  accountRulesContract: AccountRules | undefined,
  setPermissionList: (account: Permission[]) => void,
  setAccountReadOnly: (readOnly?: boolean) => void,
  filter: PermissionFilter | undefined
) => {
  if (accountRulesContract === undefined) {
    setPermissionList([]);
    setAccountReadOnly(undefined);
  } else {
    if (!filter) {
      filter = { accountAdded: null, accountGrantee: null, accountGrantor: null, blockTimestamp: null };
    }
    let eventFilter = accountRulesContract.filters.AccountAdded(
      null,
      filter.accountGrantee,
      filter.accountGrantor,
      null
    );

    accountRulesContract.queryFilter(eventFilter).then(events => {
      setPermissionList(
        events.map(e => {
          const d = new Date(e.args[3].toNumber() * 1000);

          return {
            type: 'account',
            authorizer: e.args[2],
            authorized: e.args[1],
            date: d.toLocaleString(),
            txHash: e.transactionHash,
            identifier: ''
          };
        })
      );
    });

    setAccountReadOnly(false);
  }
};

/**
 * Provider for the data context that contains the node list
 * @param {Object} props Props given to the NodeDataProvider
 * @return The provider with the following value:
 *  - nodeList: list of whiteliist enode from Node Rules contract
 *  - setNodeList: setter for the list state
 */
export const PermissionDataProvider: React.FC<{}> = props => {
  const [permissionList, setPermissionList] = useState<Permission[]>([]);
  const [accountReadOnly, setAccountReadOnly] = useState<boolean | undefined>(undefined);
  const [permissionRulesContract, setAccountRulesContract] = useState<AccountRules | undefined>(undefined);
  const [permissionFilter, setPermissionFilter] = useState<PermissionFilter>({
    accountAdded: null,
    accountGrantee: null,
    accountGrantor: null,
    blockTimestamp: null
  });

  const value = useMemo(
    () => ({
      permissionList,
      setPermissionList,
      accountReadOnly,
      setAccountReadOnly,
      permissionRulesContract,
      setAccountRulesContract,
      permissionFilter,
      setPermissionFilter
    }),
    [
      permissionList,
      setPermissionList,
      accountReadOnly,
      setAccountReadOnly,
      permissionRulesContract,
      setAccountRulesContract,
      permissionFilter,
      setPermissionFilter
    ]
  );

  const { accountIngressContract } = useNetwork();

  useEffect(() => {
    if (accountIngressContract === undefined) {
      setAccountRulesContract(undefined);
    } else {
      accountRulesFactory(accountIngressContract).then(contract => {
        setAccountRulesContract(contract);
        contract.removeAllListeners('AccountAdded');
        contract.removeAllListeners('AccountRemoved');
        contract.on('AccountAdded', () => {
          loadPermissionData(contract, setPermissionList, setAccountReadOnly, permissionFilter);
        });
        contract.on('AccountRemoved', () => {
          loadPermissionData(contract, setPermissionList, setAccountReadOnly, permissionFilter);
        });
      });
    }
  }, [accountIngressContract, permissionFilter]);

  return <DataContext.Provider value={value} {...props} />;
};

/**
 * Fetch the appropriate node data on chain and synchronize with it
 * @return {Object} Contains data of interest:
 *  - dataReady: true if isReadOnly and node allowlist are correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isReadOnly: Node contract is lock or unlock,
 *  - allowlist: list of permitted nodes from Node contract,
 */
export const usePermissionData = () => {
  function reload() {
    loadPermissionData(permissionRulesContract, setPermissionList, setAccountReadOnly, permissionFilter);
  }

  const context = useContext(DataContext);
  if (!context) {
    throw new Error('usePermissionData must be used within a PermissionDataProvider.');
  }

  const {
    permissionList,
    setPermissionList,
    accountReadOnly,
    setAccountReadOnly,
    permissionRulesContract,
    permissionFilter,
    setPermissionFilter
  } = context;

  useEffect(() => {
    loadPermissionData(permissionRulesContract, setPermissionList, setAccountReadOnly, permissionFilter);
  }, [permissionRulesContract, setPermissionList, setAccountReadOnly, setPermissionFilter, permissionFilter]);

  const formattedPermissionList = useMemo(() => {
    return permissionList
      .map(permission => ({
        ...permission,
        status: 'active'
      }))
      .reverse();
  }, [permissionList]);

  const dataReady = useMemo(() => {
    return permissionRulesContract !== undefined && accountReadOnly !== undefined && permissionList !== undefined;
  }, [permissionRulesContract, accountReadOnly, permissionList]);

  return {
    dataReady,
    allowlist: formattedPermissionList,
    isReadOnly: accountReadOnly,
    permissionRulesContract,
    setPermissionList,
    permissionFilter,
    setPermissionFilter,
    reload
  };
};
