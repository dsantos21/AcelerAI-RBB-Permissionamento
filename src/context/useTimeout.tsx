// Libs
import { useState, useEffect, useCallback } from 'react';

/**
 * Allows to clear timeouts when component unmounts
 * @return {Object} Contains:
 *  - addTimeout: Function that adds a timeoutId to be cleared
 */
export default () => {
  const [timeouts, setTimeouts] = useState<number[]>([]);

  useEffect(
    () => () => {
      timeouts.forEach(timeoutId => clearTimeout(timeoutId));
    },
    [timeouts]
  );

  const addTimeout = useCallback(timeoutId => setTimeouts(timeouts => [...timeouts, timeoutId]), [setTimeouts]);

  return { addTimeout };
};
