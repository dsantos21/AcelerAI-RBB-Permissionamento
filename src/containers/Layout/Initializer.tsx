// Libs
import React from 'react';
import PropTypes from 'prop-types';
// Context
import { useNetwork } from '../../context/network';

const Initializer: React.FC<{ NoProvider: React.FC<{}>; WrongNetwork: React.FC<{ networkId: number }> }> = ({
  children,
  NoProvider,
  WrongNetwork
}) => {
  const { networkId, isCorrectNetwork } = useNetwork();
  if (isCorrectNetwork === undefined) {
    return <NoProvider />;
  } else if (isCorrectNetwork) {
    return <div>{children}</div>;
  } else {
    return <WrongNetwork networkId={networkId!} />;
  }
};

Initializer.propTypes = {
  NoProvider: PropTypes.func.isRequired,
  WrongNetwork: PropTypes.func.isRequired
};

export default Initializer;
