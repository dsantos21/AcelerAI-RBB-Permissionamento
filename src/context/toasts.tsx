// Libs
import React, { useContext, useState, createContext, useMemo, useCallback } from 'react';
// Hooks
import useTimeout from './useTimeout';

export type Toast = {
  identifier: string;
  status: string;
  message: string;
  secondaryMessage?: string;
};

const ToastContext = createContext<{
  toasts?: Toast[];
  setToasts?: React.Dispatch<React.SetStateAction<Toast[]>>;
}>({});

/**
 * Provider for the current state of the toasts
 * @param {Object} props Props given to the ToastProvider
 * @return The provider with the following value:
 *  - toasts: Array of opened toasts
 *  - setToasts: setter of toasts
 */
export const ToastProvider: React.FC<{}> = props => {
  const [toasts, setToasts] = useState<Toast[]>([]);

  const value = useMemo(() => ({ toasts, setToasts }), [toasts, setToasts]);

  return <ToastContext.Provider value={value} {...props} />;
};

/**
 * Synchronize with the toasts state
 * @return {Object} The toasts informations and the associated setters:
 *  - toasts: Array of opened toasts
 *  - openToast: Function that creates a toast and deletes it later
 *  - updateToast: Function that updates a toast
 *  - closeToast: Function that deletes the toast
 */
export const useToast = () => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error('useToast must be used within a DataProvider.');
  }

  const { addTimeout } = useTimeout();

  const { toasts: toastsOpt, setToasts } = context;
  const toasts = toastsOpt!;

  const closeToast = useCallback(
    (targetedIdentifier: string) => () => {
      setToasts!(toasts => {
        const updatedToasts = [...toasts];
        const index = updatedToasts.findIndex(({ identifier }) => identifier === targetedIdentifier);
        updatedToasts.splice(index, 1);
        return updatedToasts;
      });
    },
    [setToasts]
  );

  const openToast = useCallback(
    (identifier: string, status: string, message: string, secondaryMessage?: string, timeout: number = 5000) => {
      const timeoutId = setTimeout(closeToast(identifier), timeout);
      closeToast(identifier)();
      setToasts!(toasts => [...toasts, { identifier, status, message, secondaryMessage }]);
      addTimeout(timeoutId);
    },
    [addTimeout, closeToast, setToasts]
  );

  const updateToast = useCallback(
    (
      targetedIdentifier: string,
      status: string,
      message: string,
      secondaryMessage?: string,
      timeout: number = 5000
    ) => {
      setToasts!(toasts => {
        const updatedToasts = [...toasts];
        const index = updatedToasts.findIndex(({ identifier }) => identifier === targetedIdentifier);
        if (index !== -1) {
          updatedToasts.splice(index, 1);
        }
        return [
          ...updatedToasts,
          {
            identifier: targetedIdentifier,
            status,
            message,
            secondaryMessage
          }
        ];
      });
      const timeoutId = setTimeout(closeToast(targetedIdentifier), timeout);
      addTimeout(timeoutId);
    },
    [addTimeout, closeToast, setToasts]
  );

  return {
    toasts,
    openToast,
    updateToast,
    closeToast
  };
};
