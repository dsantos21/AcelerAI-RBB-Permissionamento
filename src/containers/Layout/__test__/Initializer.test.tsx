// Libs
import React from 'react';
import { mocked } from 'ts-jest/utils';
import toJson from 'enzyme-to-json';
import { mount, ReactWrapper } from 'enzyme';
// Components
import Initializer from '../Initializer';

import { useNetwork } from '../../../context/network';

jest.mock('../../../context/network', () => {
  return {
    useNetwork: jest.fn()
  };
});

const NoProvider = () => <div className="noProvider" />;
const WrongNetwork = (p: { networkId: number }) => <div className="wrongNetwork" />;

describe('<Initializer />', () => {
  let wrapper: ReactWrapper<any, any, any>;

  beforeAll(() => {
    mocked(useNetwork).mockImplementation(
      () =>
        ({
          networkId: undefined,
          isCorrectNetwork: null
        } as any)
    );
  });

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = mount(
      <Initializer NoProvider={NoProvider} WrongNetwork={WrongNetwork}>
        <div className="test" />
      </Initializer>
    );
  });

  it('has called useNetwork once', () => {
    expect(useNetwork).toHaveBeenCalledTimes(1);
  });

  it('contains NoProvider={NoProvider} as a props', () => {
    expect(wrapper.props().NoProvider).toEqual(NoProvider);
  });

  it('contains WrongNetwork={WrongNetwork} as a props', () => {
    expect(wrapper.props().WrongNetwork).toEqual(WrongNetwork);
  });

  it('contains children={<div className="test" />} as a props', () => {
    expect(wrapper.props().children).toEqual(<div className="test" />);
  });

  describe('isCorrectNetwork=true, networkId=undefined', () => {
    beforeAll(() => {
      mocked(useNetwork).mockImplementation(
        () =>
          ({
            networkId: undefined,
            isCorrectNetwork: undefined
          } as any)
      );
    });

    it('does not render children when passed in', () => {
      expect(wrapper.contains(<div className="test" />)).toEqual(false);
    });

    it('contains one children', () => {
      expect(wrapper.children()).toHaveLength(1);
    });

    it('contains NoProvider', () => {
      expect(wrapper.contains(<NoProvider />)).toEqual(true);
    });

    it('matches snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('networkId=0, isCorrectNetwork=false', () => {
    beforeAll(() => {
      mocked(useNetwork).mockImplementation(
        () =>
          ({
            networkId: 0,
            isCorrectNetwork: false
          } as any)
      );
    });

    it('does not render children when passed in', () => {
      expect(wrapper.contains(<div className="test" />)).toEqual(false);
    });

    it('contains one children', () => {
      expect(wrapper.children()).toHaveLength(1);
    });

    it('contains NoProvider', () => {
      expect(wrapper.contains(<WrongNetwork networkId={0} />)).toEqual(true);
    });

    it('matches snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('networkId=0, isCorrectNetwork=true', () => {
    beforeAll(() => {
      mocked(useNetwork).mockImplementation(
        () =>
          ({
            networkId: 0,
            isCorrectNetwork: true
          } as any)
      );
    });

    it('render children when passed in', () => {
      expect(wrapper.contains(<div className="test" />)).toEqual(true);
    });

    it('matches snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
