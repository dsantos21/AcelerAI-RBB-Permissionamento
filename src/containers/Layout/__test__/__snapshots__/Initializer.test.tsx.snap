// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Initializer /> isCorrectNetwork=true, networkId=undefined matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
>
  <NoProvider>
    <div
      className="noProvider"
    />
  </NoProvider>
</Initializer>
`;

exports[`<Initializer /> networkId=0, isCorrectNetwork=false matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
>
  <WrongNetwork
    networkId={0}
  >
    <div
      className="wrongNetwork"
    />
  </WrongNetwork>
</Initializer>
`;

exports[`<Initializer /> networkId=0, isCorrectNetwork=true matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
>
  <div>
    <div
      className="test"
    />
  </div>
</Initializer>
`;
